### PR TITLE
Allows Global ReCAPTCHA #288

### PR DESCRIPTION
**Pull request by thomas-kl1**
Add the alternative domain https://www.recaptcha.net/recaptcha/api.js

### Description (*)
As a merchant I want to select the captcha domain: 

https://www.google.com/recaptcha/api.js
https://www.recaptcha.net/recaptcha/api.js
See Google Documentation: https://developers.google.com/recaptcha/docs/faq#can-i-use-recaptcha-globally

### Fixed Issues (if relevant)

1. magento/security-package#288: Allow use reCAPTCHA gloabally

### Manual testing scenarios (*)
1. ...
2. ...

### Questions or comments
Beside that the PR aims to fix components dependencies declaration.
This PR also add return type hint when possible to method signatures.

### Contribution checklist (*)
 - [ ] Author has signed the [Adobe CLA](https://opensource.adobe.com/cla.html)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

File: ReCaptchaAdminUi/Model/OptionSource.php [hunk 1]
```diff
@@ -8,6 +8,7 @@
 namespace Magento\ReCaptchaAdminUi\Model;
 
 use Magento\Framework\Data\OptionSourceInterface;
+use function array_values;
```
Comment (nathanjosiah) [inline ReCaptchaAdminUi/Model/OptionSource.php]: dont use this import

File: ReCaptchaAdminUi/Model/OptionSource.php [hunk 2]
```diff
@@ -29,14 +30,14 @@ class OptionSource implements OptionSourceInterface
      */
     public function __construct(array $options = [])
     {
-        $this->options = $options;
+        $this->options = array_values($options);
     }
 
     /**
-     * @inheritDoc
```
Comment (nathanjosiah) [inline ReCaptchaAdminUi/Model/OptionSource.php:36]: this is the correct annotation `inheritDoc`

File: ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php [hunk 3]
```diff
@@ -1,33 +0,0 @@
-<?php
```
Comment (nathanjosiah) [inline ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php:1]: Why was this deleted?

File: ReCaptchaFrontendUi/etc/frontend/di.xml [hunk 4]
```diff
@@ -11,8 +11,4 @@
                 type="Magento\ReCaptchaFrontendUi\Model\CaptchaTypeResolver"/>
     <preference for="Magento\ReCaptchaUi\Model\ErrorMessageConfigInterface"
                 type="Magento\ReCaptchaFrontendUi\Model\ErrorMessageConfig"/>
-    <type name="Magento\Framework\View\Asset\Minification">
```
Comment (nathanjosiah) [inline ReCaptchaFrontendUi/etc/frontend/di.xml:14]: Why was this removed?

File: ReCaptchaUi/Model/UiConfigResolver.php [hunk 5]
```diff
@@ -8,6 +8,10 @@
 namespace Magento\ReCaptchaUi\Model;
 
 use Magento\Framework\Exception\InputException;
+use function array_combine;
```
Comment (nathanjosiah) [inline ReCaptchaUi/Model/UiConfigResolver.php]: There's no need to import these like this.

File: ReCaptchaUi/Model/UiConfigResolver.php [hunk 6]
```diff
@@ -30,22 +36,18 @@ class UiConfigResolver implements UiConfigResolverInterface
...
-
-        foreach ($uiConfigProviders as $uiConfigProvider) {
-            if (!$uiConfigProvider instanceof UiConfigProviderInterface) {
-                throw new InputException(
-                    __('UI config provider must implement %1', [ UiConfigResolverInterface::class])
-                );
-            }
-        }
-        $this->uiConfigProviders = $uiConfigProviders;
+        $this->uiConfigProviders = array_combine(
```
Comment (nathanjosiah) [inline ReCaptchaUi/Model/UiConfigResolver.php]: I do like the functional approach but it's much less performance and much less clear what is going on here. I would switch this back.

File: ReCaptchaUi/Model/ValidationConfigResolver.php [hunk 7]
```diff
@@ -9,6 +9,9 @@
 
 use Magento\Framework\Exception\InputException;
 use Magento\ReCaptchaValidationApi\Api\Data\ValidationConfigInterface;
+use function array_combine;
```
Comment (nathanjosiah) [inline ReCaptchaUi/Model/ValidationConfigResolver.php]: No need to do this

File: ReCaptchaUi/Model/ValidationConfigResolver.php [hunk 8]
```diff
@@ -33,22 +36,18 @@ class ValidationConfigResolver implements ValidationConfigResolverInterface
...
-
-        foreach ($validationConfigProviders as $validationConfigProvider) {
-            if (!$validationConfigProvider instanceof ValidationConfigProviderInterface) {
-                throw new InputException(
-                    __('Validation config provider must implement %1.', [ConfigProviderInterface::class])
-                );
-            }
-        }
-        $this->validationConfigProviders = $validationConfigProviders;
+        $this->validationConfigProviders = array_combine(
```
Comment (nathanjosiah) [inline ReCaptchaUi/Model/ValidationConfigResolver.php]: Same with this one, let's switch it back

File: ReCaptchaUser/view/adminhtml/templates/recaptcha.phtml [hunk 9]
```diff
@@ -24,8 +24,7 @@ $isInvisible = !empty($config['invisible']);
         const scriptTag = document.getElementsByTagName('script')[0];
 
         element.async = true;
-        element.src = 'https://www.google.com/recaptcha/api.js'
-            + '?onload=globalOnRecaptchaOnLoadCallback&render=explicit';
+        element.src = '<?= /* @noEscape */ $config['api_url'] ?>?onload=globalOnRecaptchaOnLoadCallback&render=explicit';
```
Comment (nathanjosiah) [inline ReCaptchaUser/view/adminhtml/templates/recaptcha.phtml]: Just for defense in depth's sake, let's escape this with escapeJs.

**Review by nathanjosiah — CHANGES_REQUESTED** [review]

File: ReCaptchaUi/Model/ValidationConfigResolver.php (hunk 8 above)
Comment (thomas-kl1) [reply ReCaptchaUi/Model/ValidationConfigResolver.php:45]: we could get rid of this totally, as the get method is signed, we are sure that the correct type should be returned. Developers must follow the contrat of the doc anyway

File: ReCaptchaUi/Model/UiConfigResolver.php (hunk 6 above)
Comment (thomas-kl1) [reply ReCaptchaUi/Model/UiConfigResolver.php:45]: we could get rid of this totally, as the get method is signed, we are sure that the correct type should be returned. Developers must follow the contrat of the doc anyway

File: ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php (hunk 3 above)
Comment (thomas-kl1) [reply ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php:1]: https://*/recaptcha/api.js is not catchable by the Magento minifier as the url is set dynamically in js

File: ReCaptchaFrontendUi/etc/frontend/di.xml (hunk 4 above)
Comment (thomas-kl1) [reply ReCaptchaFrontendUi/etc/frontend/di.xml:14]: https://*/recaptcha/api.js is not catchable by the Magento minifier as the url is set dynamically in js

File: ReCaptchaUi/Model/ValidationConfigResolver.php (hunk 8 above)
Comment (thomas-kl1) [reply ReCaptchaUi/Model/ValidationConfigResolver.php]: I don't think we even need to have such validation. There is an implicit contract. I agree we shouldn't add noise when it has no value.

File: ReCaptchaUi/Model/UiConfigResolver.php (hunk 6 above)
Comment (thomas-kl1) [reply ReCaptchaUi/Model/UiConfigResolver.php]: I don't think we even need to have such validation. There is an implicit contract. I agree we shouldn't add noise when it has no value.

**Comment by thomas-kl1** [comment]
@magento run all tests

File: ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php (hunk 3 above)
Comment (thomas-kl1) [reply ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php:1]: @nathanjosiah

File: ReCaptchaFrontendUi/etc/frontend/di.xml (hunk 4 above)
Comment (thomas-kl1) [reply ReCaptchaFrontendUi/etc/frontend/di.xml:14]: @nathanjosiah

File: ReCaptchaUi/Model/UiConfigResolver.php (hunk 6 above)
Comment (thomas-kl1) [reply ReCaptchaUi/Model/UiConfigResolver.php]: @nathanjosiah

File: ReCaptchaUi/Model/ValidationConfigResolver.php (hunk 8 above)
Comment (thomas-kl1) [reply ReCaptchaUi/Model/ValidationConfigResolver.php]: @nathanjosiah

**Comment by thomas-kl1** [comment]
hi @nathanjosiah could you take a look on my comments?

**Comment by thomas-kl1** [comment]
poke @nathanjosiah

File: ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php (hunk 3 above)
Comment (nathanjosiah) [reply ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php:1]: My concern is that this will break existing expectations for minification exclusion. Are you saying this was not working already before your changes? As far as I can see the URL was static before you added the new URL.

**Comment by nathanjosiah** [comment]
@thomas-kl1 Apologies for the delay, I am on paternity leave but I left a comment and resolved some existing ones.

**Comment by thomas-kl1** [comment]
@nathanjosiah ho! Congrats! :confetti_ball: :confetti_ball:

File: ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php (hunk 3 above)
Comment (thomas-kl1) [reply ReCaptchaFrontendUi/Plugin/ExcludeFromMinification.php:1]: I'm not sure to understand, but the only static url in the template was ReCaptchaUser/view/adminhtml/templates/recaptcha.phtml and it's removed with this PR. So I thought it's safe to remove it from the minification exclusion list.

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
The failures in tests does not seems to be related to this PR.

**Comment by nathanjosiah** [comment]
Connected to #288

**Comment by nathanjosiah** [comment]
@thomas-kl1 can you merge develop and rerun builds?

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by nathanjosiah** [comment]
@thomas-kl1 Something appears to be broken with MFTF test generation. The error log indicates it is unable to generate at least one of the tests. Can you check that you are able to build locally?

**Comment by thomas-kl1** [comment]
Hi @nathanjosiah Would be easier for me if you could actually give the result of mftf.log, it's really hard to run test locally with the complete suite..

**Comment by nathanjosiah** [comment]
@magento run Functional Tests CE  against 2.4-develop

**Comment by nathanjosiah** [comment]
@magento run Functional Tests CE against 2.4.3-develop

**Comment by nathanjosiah** [comment]
@thomas-kl1 Looks like there are some errors in certain scenarios in the latest build (see https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/security-package/pull/289/bf46e4fe54b374d1b9fcff72e9c97fce/Functional/allure-report-ce/index.html#categories/bda2d1d522cd5c6d3b119ea9eba8bd02/6f6a43bc714a93f9/) specifically
```
Magento_ReCaptchaFrontendUi/js/reCaptcha.js 53:68 Uncaught TypeError: Cannot read property 'api_url' of undefined
```

**Comment by nathanjosiah** [comment]
@thomas-kl1 Pinging again for visibility.

**Comment by thomas-kl1** [comment]
Hi @nathanjosiah I'll rebase and take a look when I'll have some free time, sorry for the delay

**Comment by nathanjosiah** [comment]
No worries. I wanted to make sure you saw my comment and that you were
still wanting to work on this issue

**Comment by nathanjosiah** [comment]
Closing due to inactivity.

**Comment by thomas-kl1** [comment]
Rly? Make it draft but don't close it
